dts/bindings/regulator/regulator-gpio.yaml

```diff
@@ -0,0 +1,96 @@
+# Copyright 2023 EPAM Systems
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  GPIO-controlled voltage of regulators
+
+  Example of dts node:
+    vccq_sd0: regulator-vccq-sd0 {
+      compatible = "regulator-gpio";
+
+      regulator-name = "SD0 VccQ";
+      regulator-min-microvolt = <1800000>;
+      regulator-max-microvolt = <3300000>;
+
+      enable-gpios = <&gpio5 3 GPIO_ACTIVE_HIGH>;
+
+      gpios = <&gpio5 1 GPIO_ACTIVE_HIGH>, <&gpio5 2 GPIO_ACTIVE_HIGH>;
+      gpios-states = <1>;
+      states = <3300000 2>, <2700000 1>, <1800000 0>;
+
```

Comment:
Thanks for updating the description.
Have you experimented with using the linear_range API?

e.g;
Define states as a 1D array:
`states = <1800000>, <2700000>, <3300000>`
 
In the initialisation macros, convert this into:

```
static const struct linear_range volt_ranges[] = {
	LINEAR_RANGE_INIT(1800000, 0, 0U, 0U), 
	LINEAR_RANGE_INIT(2700000, 0, 0U, 0U),
	LINEAR_RANGE_INIT(3300000, 0, 0U, 0U)};
```

Then you could use the linear_range API for your get/set/list voltage commands which would simplify that code a great deal.

Reply:
First of all, sorry for the delay.

I've reviewed the Linear Range functionality and I don't think it is suitable for this driver.

* On one hand, during the review, there was a request to align bindings and DTS with the Linux Kernel. However, here I see an opposite request:
> Define states as a 1D array:
> states = <1800000>, <2700000>, <3300000>

* potentially, the current implementation can change more GPIO states, such as 32 GPIOs. I'm not sure if we need that many, but in the case where we have a zero step in the Linear Range, `min_idx` is the only value used, and it has only 16 bits. This is because I'm using `idx` as a bit-array for GPIO states. Therefore, I need to change the type of idx and possibly modify related code in the Linear Range API.
* it is necessary to ensure that the user fills in the states in a sorted manner. Otherwise, during looping through the ranges, I may set a non-close value to the requested one. For instance, if the states are "states = <1800000>, <10000000>, <3300000>" and I request a voltage between 3.1V and 3.4V, it may return an index for 10V. You can see how `linear_range_group_get_win_index` or `linear_range_group_get_index` works in this case. Therefore, there are three possible solutions in this case, and all of them add some extra complexity: change the Linear Range API, sort the array during the driver's initialization, or add checks during initialization and write a note to the user that the array needs to be fixed.
* the complexity of the code has increased slightly for the set function because we need to go through two loops, one for ranges and one inside the range. Rest of functionality relating to setting of GPIOs states remains the same.

I'm not sure if the code, after all the changes, will be simpler than it is now. However, if you really want me to remake this implementation, I'll do it. Just let me know.

Reply:
Thanks for looking into this.  Given the complexity of changing I am happy with the implementation :-)